Block decoding support.

vext01 opened the pull request:
This adds support for block decoding into the perf_pt backend.

It's a large change and I'm pretty certain you will be able to find some nits! Let the discussion begin!

Anything `XXX` is either Rust/Cargo bugs/limitations, or something which I think can be tackled later in later PRs.

Fixes #28

vext01 commented on src/lib.rs:
```diff
@@ -56,8 +103,12 @@ pub trait Trace: Debug {
+    fn to_file(&self, file: &mut File);
+
+    /// Iterate over the blocks of the trace.
+    /// XXX if this were to return a `Result` then the `Block` type could disappear?
```
This is what we discussed on the walk in to King's today. I think this can be the next PR.

vext01 commented on src/backends/perf_pt/mod.rs:
```diff
@@ -334,11 +419,132 @@ impl Drop for PerfPTTracer {
+        Some((tracer, trace, expects.unwrap()))
+    }
 
+    // XXX This isn't very flexible, as it only allows us to call test helpers with that one
```
And this will probably be the 2nd in line for future PRs.

vext01 commented:
Unrelated network error on travis. kicked.

ltratt commented on build.rs:
```diff
@@ -80,6 +80,22 @@ fn build_libipt() {
 }
 
+fn fetch_libipt() {
+    eprintln!("Fetch libipt...");
```
Shouldn't this be "Build libipt" (and do we need the `...`?).

vext01 replied:
There are separate build and fetch steps as we need to build a couple of files inside the libipt repo regardless of whether the user wants to use their own libipt or ours.

There's a comment later that tries to explain, I can move it up if you think it would be clearer.

vext01 replied:
`...` could go, yeah.

ltratt replied:
Ah, yes, I see now. Moving the comment might help.

ltratt commented on build.rs:
```diff
@@ -115,11 +147,14 @@ fn main() {
         }
     }
     c_build.file("src/util/util.c");
+
```
Extraneous blank line?

vext01 replied:
It was intentional, as it splits configuring the C build from actually doing it. Can kill it if you like.

ltratt commented on examples/simple_example.rs:
```diff
@@ -65,36 +64,53 @@ fn tracer() -> Box<Tracer> {
 
+/// Prints the addresses of the first `qty` blocks in a trace along with it's name and
+/// computation result.
+fn print_trace(trace: &Box<Trace>, name: &str, result: u32, qty: usize) {
```
This feels like a debug function to me? And should it be printing directly or building a string and returning it to the user? It feels a bit odd to pass a `name` to a function that prints it straight out to stdout.

ltratt replied:
Doh, ignore the "debug" comment -- I missed that this as an example :/

vext01 replied:
Yeah, it's an example. Just doing something with the trace to offer the user some evidence that we actually did trace something :)

ltratt commented on examples/simple_example.rs:
```diff
@@ -65,36 +64,53 @@ fn tracer() -> Box<Tracer> {
+    println!("{}: num_blocks={}, result={}", name, count, result);
+    let mut itr = trace.iter_blocks();
+    let mut end = false;
+    for i in 0..qty {
```
I think you can simplify most of the code in this function down to something like:

```rust
trace.iter_blocks().take(qty).map(|x| println!("  block ...", ...));
if qty > count {
  println!("  ... {} more", ...);
}
```

vext01 replied:
Sounds good.

vext01 replied:
Is there a better way than using `last()` to consume the iterator?

vext01 replied:
For context:
```
warning: unused `std::iter::Map` which must be used: iterator adaptors are lazy and do nothing unless consumed
  --> examples/simple_example.rs:73:5
   |
73 | /     trace.iter_blocks().take(qty)
74 | |                        .enumerate()
75 | |                        .map(|(i, blk)| println!("  block {}: 0x{:x}", i, blk.unwrap().start_vaddr()));
   | |______________________________________________________________________________________________________^
   |
```

ltratt replied:
Good point. This should probably be a `for` loops in retrospect:

```rust
for x in trace.iter_blocks().take(qty) {
  // do something with x
}
```

vext01 replied:
Using `.last()` works fine. I think with a comment I actually prefer what you were proposing.

```diff
diff --git a/examples/simple_example.rs b/examples/simple_example.rs
index 2260d51..6b083a9 100644
--- a/examples/simple_example.rs
+++ b/examples/simple_example.rs
@@ -69,28 +69,24 @@ fn tracer() -> Box<Tracer> {
 fn print_trace(trace: &Box<Trace>, name: &str, result: u32, qty: usize) {
     let count = trace.iter_blocks().count();
     println!("{}: num_blocks={}, result={}", name, count, result);
-    let mut itr = trace.iter_blocks();
-    let mut end = false;
-    for i in 0..qty {
-        if let Some(blk) = itr.next() {
-            println!("  block {}:\t0x{:x}", i, blk.unwrap().start_vaddr());
-        } else {
-            end = true;
-        }
-    }
-    if !end {
+
+    trace.iter_blocks().take(qty)
+                       .enumerate()
+                       .map(|(i, blk)| println!("  block {}: 0x{:x}", i, blk.unwrap().start_vaddr()))
+                       .last();
+    if count > qty {
         println!("  ... {} more", count - qty);
     }
     println!("");
 }
```

ltratt replied:
The Rust docs say (and I was being an idiot for not thinking of this) "The idiomatic way to write a map for its side effects is to use a for loop instead".

vext01 replied:
Fair enough. I'll see what I can come up with.

ltratt commented on examples/simple_example.rs:
```diff
@@ -65,36 +64,53 @@ fn tracer() -> Box<Tracer> {
+}
+
+/// It's up to you to ensure the compiler doesn't move or optimise out the computation you intend
+/// to trace. Here we've used an uninlinable function. See also `test::black`.
```
`test::black_box`?

vext01 replied:
oops, yes.

ltratt commented on examples/simple_example.rs:
```diff
@@ -65,36 +64,53 @@ fn tracer() -> Box<Tracer> {
+fn work() -> u32 {
+    let mut res = 0;
+    for _ in 0..50 {
+        // Computation which blocks the compiler from eliminating the loop.
```
s/blocks/stops/

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+ *
+ * `vdso_fd` is an open file descriptor for the filename `vdso_filename`. This
+ * is where the VDSO code will be written. libipt will read this file lazily,
+ * so it's up to the caller to make sure this file lives long enough for their
```
Does "the caller" mean "us"?

vext01 replied:
The caller of this function, which is "us" in a sense. This detail is not exposed to the consumers of our library.

ltratt replied:
Right, so I think being clearer that it's "us" will help.

vext01 replied:
Hrm. Actually, I think that's clear from the source. Look at the sole call site of this function:
https://github.com/softdevteam/hwtracer/blob/ffa6ee0a521aa5f201d446a2c4aff4a979b3e3cb/src/backends/perf_pt/mod.rs#L87-L99

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+ *
+ * Returns a pointer to a configured libipt block decoder or NULL on error.
+ *
+ * XXX This function assumes that we are decoding a trace for the current CPU
```
This doesn't feel like an XXX to me: it's a fundamental assumption of the whole library, right?

vext01 replied:
The CPU part is should probably be baked in.

With regards to "the current process", I don't see why our library couldn't be used more generally to trace other processes. We even offer a configuration option to choose an arbitrary `tid` to trace. But if we wanted to decode a trace for another process, we would have to offer an API to load the code for a process other than ourself.

ltratt replied:
Maybe someone might do that as a PR, but I can't see that we'd ever want this?

vext01 replied:
Here's the question: Is hwtracer a library for our meta-tracer, or is it general purpose?

ltratt replied:
I am not generally a fan of putting in functionality for use cases that we don't know will ever exist. So I think, until someone gives us a use case that suggests that we should generalise, we might as well keep things simple.

vext01 replied:
Fair enough. FWIW, we have #49 to remind us that this is on the cards.

ltratt replied:
I'm not sure having that as an issue is very useful (*we're* never going to do it AFAICS), but up to you.

vext01 replied:
I think the correct thing to do is to update the README and clearly state that the library is currently for tracing one's self.

ltratt replied:
I agree.

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+    }
+
+    // Work around CPU bugs.
+    if (config.cpu.vendor) {
```
Use `&&` for the nested `if`?

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+    if (*decoder_status < 0) {
+        // Error.
+        return false;
+    } else if (*decoder_status & pts_eos) {
```
What's `pts_eos`? I'm not sure I understand the second and third branches of this `if` statement TBH.

vext01 replied:
`pts_eos` means "PT status: end of stream". In other words, there are no more blocks to iterate over. We return 0 which prompts the Rust iterator to return `None`.

The third case handles statuses that I've not yet considered. There's a list of possibilities here:
https://github.com/01org/processor-trace/blob/24982c1a6fce48f1e416461d42899805f74fbb26/libipt/include/intel-pt.h.in#L1064-L1074

* `pts_event_pending` should not occur, since this very call, by definition, should be draining the events from the stream.
* I don't think `pts_ip_suppressed` can happen as this call doesn't fetch any blocks.

None the less, I'd like to know if we get here. Would it be better as an assertion?

ltratt replied:
Yes, maybe an `assert` would be better because at least we'd know if they *can* happen.

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+    }
+
+    if (block.ninsn == 0) {
+        // A block with zero instructions. Shouldn't happen.
```
If it shouldn't happen, this should be an `assert`?

vext01 replied:
It could be yeah. Do you think that's better than returning an error?

ltratt replied:
If it's not supposed to happen, then I think so.

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+load_self_image(struct load_self_image_args *args)
+{
+    bool ret = dl_iterate_phdr(load_self_image_cb, args) == 0;
+    if (fsync(args->vdso_fd) == -1) {
```
Can't we simplify this to `return fsync(...) != -1;`?

vext01 replied:
We can!

vext01 replied:
Wait, what if `dl_iterate_phdr` fails, but fsync succeeds? We would return `true` ignoring the error.

ltratt replied:
That's a good point. But should we do `fsync` if `dl_iterate_phdr` failed? Maybe we should really be doing something like:

```
return !((dl_iterate_phdr(load_self_image_cb, args) == 0) && (fsync(args->vdso_fd) != -1));
```

vext01 replied:
We can do, but after splitting the long line, it's probably not going to be pretty. I'll try it nonetheless.

ltratt replied:
Maybe just two `if`s then.

vext01 replied:
It didn't look too bad in the end.

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+    ElfW(Phdr) phdr;
+    ElfW(Half) i;
+
+    (void) size;
```
I'm not sure what this line is doing.

vext01 replied:
It's a standard way to silence the compiler in the event of an unused argument. The argument is forced upon us by the signature of the `dl_iterate_phdr` callback.

ltratt replied:
Aha! Maybe doc that with `// Unused variable: silence compiler warnings` or similar?

vext01 replied:
I don't think we need to be quite that verbose, but I agree.

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+bool
+dump_vdso(int fd, uint64_t vaddr, size_t len)
+{
+    size_t wrote_so_far = 0;
```
If you want a long variable name, how about `written`?

vext01 replied:
I like it!

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -35,13 +35,317 @@
+        }
+        wrote_so_far += wrote;
+    }
+
```
I think all of the other functions up to this point haven't had a blank line before `return`.

vext01 replied:
Just ha d a look. It's currently a mix.

ltratt commented on src/backends/perf_pt/test_helpers.c:
```diff
@@ -0,0 +1,144 @@
+bool
+perf_pt_append_self_ptxed_raw_args(struct append_args *rs_args)
+{
+    if (dl_iterate_phdr(append_self_ptxed_raw_args_cb, rs_args) == 1) {
```
Can just be `return dl_iterate_phdr(...) != 1;`

vext01 replied:
Yeah. Don't know why I made that so complicated. It probably had some other code around it at some point.

ltratt commented on src/backends/perf_pt/test_helpers.c:
```diff
@@ -0,0 +1,144 @@
+ * `data` is a pointer to a Rust `AppendSelfPtxedArgs` instance.
+ */
+static int
+append_self_ptxed_raw_args_cb(struct dl_phdr_info *info, size_t size, void *data)
```
Most of this function is the same as `load_self_image_cb` -- can/should we factor out the common bits?

vext01 replied:
I did consider it, but thought it might be simpler to comprehend by having them separate like this. How strongly do you feel about it?

It might be a bit weird/confusing to have `void *data` be either a pointer to a Rust struct or a C `pt_image`.

ltratt replied:
It's always a judgement call as to when to copy/paste and when to factor out. Your decision (I just wanted to flag it).

vext01 replied:
I think in this case it would be best to leave them apart for the reason above.

ltratt replied:
OK.

ltratt commented on src/lib.rs:
```diff
@@ -94,31 +145,40 @@ enum TracerState {
+    pub fn work_loop(iters: u64) -> u64 {
+        let mut res = 0;
+        for _ in 0..iters {
+            // Computation which blocks the compiler from eliminating the loop.
```
s/blocks/stops/

vext01 commented:
The last commit addresses your comments.

I think I'd like to rename the functions in `decode.c` so as to make clear we are decoding blocks (not instructions, which libipt supports too). OK by you?

vext01 commented:
Like this ^

ltratt commented on examples/simple_example.rs:
```diff
@@ -69,28 +69,23 @@ fn tracer() -> Box<Tracer> {
-            end = true;
-        }
+
+    for (i, blk) in trace.iter_blocks().take(qty) .enumerate() {
```
Stray space before `.enumerate`.

ltratt commented on src/backends/perf_pt/decode.c:
```diff
@@ -106,11 +105,9 @@ perf_pt_start_decode(void *buf, uint64_t len, int vdso_fd, char *vdso_filename,
-            failing = true;
-            goto clean;
-        }
+    if ((config.cpu.vendor) && (pt_cpu_errata(&config.errata, &config.cpu) < 0)) {
```
No need for brackets around `config.cpu.vendor`.

ltratt commented:
>I think I'd like to rename the functions in decode.c so as to make clear we are decoding blocks (not instructions, which libipt supports too). OK by you?

Fine with me.

vext01 commented:
So is it squash time?

ltratt commented:
Yep!

vext01 commented:
splat!
